transport/dnstruncate/packet_listener.go

```diff
@@ -0,0 +1,252 @@
+	addr net.Addr        // DNS resolver's network address
+	n    int             // Indicates how many bytes have been processed by ReadFrom
+	wg   *sync.WaitGroup // To mark that ReadFrom has finished processing this request
+}
+
+// dnsTruncateConn is an implementation of [net.PacketConn] that handles DNS requests in UDP protocol, without sending
+// the requests to the actual DNS resolver. It sets the TC (truncated) bit in the DNS response header to tell the
+// caller to resend the DNS request over TCP.
+//
+// dnsTruncateConn uses a single channel architecture to bridge DNS requests from WriteTo and DNS handling logic in
+// ReadFrom. There is no need to allocate any additional buffers; and all methods are thread-safe.
+type dnsTruncateConn struct {
+	done chan struct{}
+	req  chan *dnsReq
+
+	rdDdl *ddltimer.DeadlineTimer
+	wrDdl *ddltimer.DeadlineTimer
+}
+
+// Compilation guard against interface implementation
+var _ transport.PacketListener = (*dnsTruncateListener)(nil)
+var _ net.PacketConn = (*dnsTruncateConn)(nil)
+
+// NewPacketListener creates a new [transport.PacketListener] that can be used to handle DNS requests if the remote
+// proxy doesn't support UDP traffic. It sets the TC (truncated) bit in the DNS response header to tell the caller to
+// resend the DNS request over TCP.
+//
+// This [transport.PacketListener] should only be used if the remote proxy server doesn't support UDP traffic at all.
+// Note that all other non-DNS UDP packets will be dropped by this [transport.PacketListener].
+func NewPacketListener() (transport.PacketListener, error) {
+	return &dnsTruncateListener{}, nil
+}
+
+// ListenPacket creates a new [net.PacketConn] to handle UDP traffic.
+func (h *dnsTruncateListener) ListenPacket(ctx context.Context) (net.PacketConn, error) {
+	conn := &dnsTruncateConn{
+		done:  make(chan struct{}),
+		req:   make(chan *dnsReq),
+		rdDdl: ddltimer.New(),
+		wrDdl: ddltimer.New(),
+	}
+	return conn, nil
+}
+
+// Close closes the connection. Any blocked ReadFrom or WriteTo operations will be unblocked.
+func (c *dnsTruncateConn) Close() error {
+	select {
+	case <-c.done:
+		// prevent panic of closing a closed channel
+	default:
+		close(c.done)
+	}
+	c.rdDdl.Stop()
+	c.wrDdl.Stop()
+	return nil
+}
+
+// ReadFrom reads a DNS request packet from a pending WriteTo, setting the TC (truncated) bit and copying the response
+// into p. It returns the number of bytes copied into p and the DNS resolver's address (passed to WriteTo).
+//
+// ReadFrom returns [io.EOF] if the connection has been closed. It returns [os.ErrDeadlineExceeded] if the read
+// deadline is exceeded; see SetDeadline and SetReadDeadline.
+//
+// ReadFrom will block until one WriteTo sends a DNS request. So you need to use goroutine to send and receive the
+// response:
+//
+//	go c.WriteTo(request, resolverAddr)
+//	c.ReadFrom(response)
+func (c *dnsTruncateConn) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
```

Comment:
We are jumping over unnecessary hoops here. That's because the interface we are offering on `lwip2transport.Device` turned out to be inconvenient. It forces us to implement a packet listener, so it can create a `PacketConn` and then relay.

A packet proxy handler would be a much easier to implement interface.

The Packet proxy handler could have an interface like
```go
type HandleRequest func(packet []byte, dstAddr string) (int, error)

type PacketProxyHandler interface {
    Proxy(responseWriter WriteFromCloser) (HandleRequest, error)
}
```
In that case, `Proxy` doesn't need to set up a packet listener, and the `HandleRequest` function returned would write the truncated response to the `responseWriter`

This is a simpler implementation and would remove the need for a [go routine per socket](https://github.com/Jigsaw-Code/outline-internal-sdk/blob/33ceec5625c075363bebb0979d9818cce39eee8c/network/lwip2transport/udp.go#L66).

That will require some changes in the `lwip2transport.Device`.

For the normal handler case, `Proxy()` would create a `PacketConn` with a listening socket, and set up a relay to read from the `PacketConn` and write to the responseWriter. It would return a `HandleRequest` function that writes to the `PacketConn`.

We should provide an adaptor handler that takes the PacketListener and converts to a PacketProxyHandler.

I believe that such change would make it easier for others to provide custom handlers.

Thoughts?

Reply:
refactored in #27